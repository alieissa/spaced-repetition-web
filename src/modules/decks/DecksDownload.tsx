/* eslint-disable jsx-a11y/anchor-has-content */

import { useEffect, useRef } from 'react'
import 'src/App.css'
import { DownloadButton } from '../../components'
import { useDownloadDecks } from './decks.hooks'
/**
 * This contains the download button. Once the button is clicked all the users
 * decks are imported in a file called decks.json
 */
export default function DecksDownload() {
  const downloadAnchorRef = useRef<HTMLAnchorElement>(null)
  const [downloadDecksUrl, downloadDecksStatus, downloadDecks] =
    useDownloadDecks()

  const status = downloadDecksStatus.type
  const isDownloadDecksDisbaled = status === 'Loading' || status === 'Failure'

  useEffect(() => {
    if (status !== 'Success') {
      return
    }
    downloadAnchorRef.current?.click()
  }, [status])

  // TODO display success or failure message component on page wher
  // download button is displayed
  return (
    <>
      <DownloadButton
        data-testid="decks-download-btn"
        disabled={isDownloadDecksDisbaled}
        onClick={downloadDecks}
      />

      {downloadDecksUrl && (
        <a
          hidden
          download
          data-testid="decks-download-anchor"
          ref={downloadAnchorRef}
          href={downloadDecksUrl}
        />
      )}
    </>
  )
}
