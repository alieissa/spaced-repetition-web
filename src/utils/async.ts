/** @format */
/**
 * Credit goes to https://medium.com/@fillopeter/pattern-matching-with-typescript-done-right-94049ddd671c
 * For the pattern matching code
 */
export type Untriggered = {
  type: 'Untriggered'
}
export function Untriggered(): Untriggered {
  return {
    type: 'Untriggered',
  }
}

const t: Optional<Answer, 'easiness' | 'quality' | 'interval'> = {
  id: '',
  content: 'fsdfds',
}
export type Loading<V> = {
  type: 'Loading'
  value: V
}
export function Loading<V>(v: V): Loading<V> {
  return {
    type: 'Loading',
    value: v,
  }
}

export type Failure<E> = {
  type: 'Failure'
  value: E
}
export function Failure<E>(e: E): Failure<E> {
  return {
    type: 'Failure',
    value: e,
  }
}

export type Success<V> = {
  type: 'Success'
  value: V
}
export function Success<V>(v: V): Success<V> {
  return {
    type: 'Success',
    value: v,
  }
}

export type Async<L = unknown, F = unknown, S = unknown> =
  | Untriggered
  | Loading<L>
  | Failure<F>
  | Success<S>

export type AsyncStatus<R, E, A> = Async<R, E, A>['type']

export type AsyncMap<R, E, A, U> = {
  [K in AsyncStatus<R, E, A>]: U extends { type: K } ? U : never
}

export type AsyncStatusMap<R, E, A> = AsyncMap<R, E, A, Async<R, E, A>>
export type AsyncPattern<T, R, E, A> = {
  [K in keyof AsyncStatusMap<R, E, A>]: (a: AsyncStatusMap<R, E, A>[K]) => T
}

export function match<R, E, A>(
  e: Async<R, E, A>,
): <T>(pattern: AsyncPattern<T, R, E, A>) => T {
  return <T>(pattern: AsyncPattern<T, R, E, A>) => pattern[e.type](e as any)
}
