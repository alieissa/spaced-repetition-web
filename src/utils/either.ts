/** @format */

/**
 * Credit goes to https://medium.com/@fillopeter/pattern-matching-with-typescript-done-right-94049ddd671c
 * For the pattern matching code
 */
export type Right<V> = Readonly<{
  type: 'Right'
  value: V
}>
export function Right<V>(v: V): Right<V> {
  return {
    type: 'Right',
    value: v,
  }
}

export type Left<E = Error> = Readonly<{
  type: 'Left'
  value: E
}>
export function Left<E>(e: E): Left<E> {
  return {
    type: 'Left',
    value: e,
  }
}

export type Either<L, R> = Left<L> | Right<R>

export type EitherType<L, R> = Either<L, R>['type']
export type EitherMap<L, R, U> = {
  [K in EitherType<L, R>]: U extends { type: K } ? U : never
}
export type EitherTypeMap<L, R> = EitherMap<L, R, Either<L, R>>
export type EitherPattern<T, L, R> = {
  [K in keyof EitherTypeMap<L, R>]: (either: EitherTypeMap<L, R>[K]) => T
}

export function match<L, R>(
  e: Either<L, R>,
): <T>(pattern: EitherPattern<T, L, R>) => T {
  return <T>(pattern: EitherPattern<T, L, R>) => pattern[e.type](e as any)
}
