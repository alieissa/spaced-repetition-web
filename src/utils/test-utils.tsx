/** @format */

import { combineReducers, configureStore } from '@reduxjs/toolkit'
import { render } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import { RouterProvider, createMemoryRouter } from 'react-router-dom'
import decksReducer from 'src/modules/decks/decks.reducer'

// As a basic setup, import your same slice reducers
import { loginReducer } from 'src/modules/login'
import { signupReducer } from 'src/modules/signup'
import { verificationReducer } from 'src/modules/verification'

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
// interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
//   preloadedState?: PreloadedState<RootState>
//   store?: AppStore
// }

const reducers = combineReducers({
  login: loginReducer,
  signup: signupReducer,
  verification: verificationReducer,
  decks: decksReducer,
})

const withRedux = (ui: React.ReactElement) => {
  const store = configureStore({ reducer: reducers })
  return <Provider store={store}>{ui}</Provider>
}
const withRouter = (ui: React.ReactElement) => {
  const router = createMemoryRouter([{ element: ui, path: '/' }], {
    initialEntries: ['/'],
    initialIndex: 1,
  })

  return <RouterProvider router={router} />
}
export function renderWithProviders(ui: React.ReactElement) {
  // Return an object with the store and all of RTL's query functions
  return render(withRedux(withRouter(ui)))
}

export const flushPromises = () =>
  new Promise((resolve) => setTimeout(resolve, 0))
