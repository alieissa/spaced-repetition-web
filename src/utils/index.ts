/** @format */

import * as async from './async'
import * as either from './either'

export { async }
export { either }
