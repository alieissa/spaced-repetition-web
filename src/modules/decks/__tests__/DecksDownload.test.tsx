/** @format */

import { faker } from '@faker-js/faker'
import '@testing-library/jest-dom'
import { act, screen } from '@testing-library/react'
import user from '@testing-library/user-event'
import { rest } from 'msw'
import { setupServer } from 'msw/lib/node'
import { flushPromises, renderWithProviders } from 'src/utils/test-utils'
import DecksDownload from '../DecksDownload'

const decksDownloadUrl = `${process.env.REACT_APP_API_ENDPOINT}/decks/download`

const server = setupServer()
beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

describe('DecksDownload', () => {
  describe('view', () => {
    it('should render correctly', () => {
      const { asFragment } = renderWithProviders(<DecksDownload />)
      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe('interaction', () => {
    it('should disable download button when downloading', async () => {
      // Assemble
      server.use(
        rest.get(decksDownloadUrl, async (__, res, ctx) => {
          return res(ctx.status(200), ctx.delay(100))
        }),
      )
      renderWithProviders(<DecksDownload />)
      const decksDownloadBtn = screen.getByTestId('decks-download-btn')

      // Act
      await act(() => user.click(decksDownloadBtn))

      // Assert
      expect(decksDownloadBtn).toBeDisabled()
    })

    it('should disable download button when download failed', async () => {
      // Assemble
      server.use(
        rest.get(decksDownloadUrl, (__, res, ctx) => res(ctx.status(422))),
      )
      renderWithProviders(<DecksDownload />)
      const decksDownloadBtn = screen.getByTestId('decks-download-btn')

      // Act
      await act(() => user.click(decksDownloadBtn))
      await act(flushPromises)

      // Assert
      expect(decksDownloadBtn).toBeDisabled()
    })

    it('should add download url to download anchor', async () => {
      // Assemble
      const anchorHref = faker.internet.url()
      faker.internet.url()
      server.use(
        rest.get(decksDownloadUrl, async (__, res, ctx) => {
          return res(ctx.status(200), ctx.text(anchorHref))
        }),
      ),
        renderWithProviders(<DecksDownload />)
      const decksDownloadBtn = screen.getByTestId('decks-download-btn')

      // Act
      await act(() => user.click(decksDownloadBtn))
      await act(flushPromises)

      // Assert
      const decksDownloadAnchor = screen.getByTestId('decks-download-anchor')
      expect(decksDownloadAnchor).toHaveAttribute('href', anchorHref)
    })
  })
})
