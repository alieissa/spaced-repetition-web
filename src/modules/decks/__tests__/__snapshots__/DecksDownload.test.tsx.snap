// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DecksDownload view should render correctly 1`] = `
<DocumentFragment>
  <button
    class="ui small icon button export-btn"
    data-testid="decks-download-btn"
  >
    <i
      aria-hidden="true"
      class="download icon"
    />
  </button>
</DocumentFragment>
`;
